core/templates/domain/exploration/RuleInputTypeFactory.ts

```diff
@@ -0,0 +1,158 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory to determine correct type of RuleInputs
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { IRuleInput } from 'domain/exploration/RuleObjectFactory';
+import { INumberWithUnitsBackendDict } from
+  'domain/objects/NumberWithUnitsObjectFactory';
+import { IFractionDict } from
+  'domain/objects/FractionObjectFactory.ts';
+/* eslint-disable max-len */
+import { INote } from 'extensions/interactions/MusicNotesInput/directives/music-notes-input-rules.service';
+/* eslint-enable max-len */
+import { IGraphBackendDict } from
+  'extensions/interactions/GraphInput/directives/graph-detail.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class RuleInputTypeFactory {
+  _isGraph(variable: Object): variable is IGraphBackendDict {
+    if (this._isString(variable) || this._isNumber(variable)) {
+      return false;
+    }
+
+    return 'vertices' in variable;
+  }
+
+  _isNoteArray(variable: Object): variable is INote[] {
+    if (Array.isArray(variable)) {
+      if (variable.length === 0) {
+        return true;
+      }
+
+      if (typeof variable[0] === 'string' || typeof variable[0] === 'number') {
+        return false;
+      }
+
+      return 'readableNoteName' in variable[0];
+    }
+
+    return false;
+  }
+
+  _isNumberWithUnits(
+      variable: Object): variable is INumberWithUnitsBackendDict {
+    if (this._isString(variable) || this._isNumber(variable)) {
+      return false;
+    }
+
+    return 'fraction' in variable;
+  }
+
+  _isFraction(variable: Object): variable is IFractionDict {
+    if (this._isString(variable) || this._isNumber(variable)) {
+      return false;
+    }
+
+    return 'numerator' in variable;
+  }
+
+  _isString(variable: Object): variable is string {
+    return typeof variable === 'string';
+  }
+
+  _isNumber(variable: Object): variable is number {
+    return typeof variable === 'number';
+  }
+
+  _isStringArray(variable: Object): variable is string[] {
+    return (Array.isArray(variable) && (
+      variable.length === 0 || typeof variable[0] === 'string'));
+  }
+
+  _isStringArrayArray(variable: Object): variable is string[][] {
+    return (Array.isArray(variable) && (
+      variable.length === 0 || (Array.isArray(variable[0]) && (
+        variable[0].length === 0 || typeof variable[0][0] === 'string'))));
+  }
+
+  _isNumberArray(variable: Object): variable is number[] {
+    return (Array.isArray(variable) && (
+      variable.length === 0 || typeof variable[0] === 'number'));
+  }
+
+  graphInstance(variable: IRuleInput): IGraphBackendDict {
```

Comment:
Maybe rename to smtg like `verifyGraphInstance` since it is difficult to understand what the function does with the current name.

Also, what happens if it is not an instance, shouldn't  there be a return null or maybe throw Error?

Ditto below

Reply:
1. Okay, I think `getGraphInstance` seems more appropriate. WDYT?
2. Yeah, right. done! (Thrown an error if it is not an instance).
3. Same variable is returned because we just using it change the type of `variable` because want the `ts compiler` to know that variable is of type  `IGraphBackendDict` (or other).
4. It could be, but this way we can assure that only `IRuleInput` is passed as an argument.

Reply:
Ah ok, so you basically type casting the variable from an IRuleInput to IGraphBackendDIct for ts compiler. Is that it? Add that in a comment for this function.
Knowing this, yeah, `getGraphInstance` sounds reasonable.

Reply:
done!